text/1015-full-text-search.rst

```diff
@@ -73,59 +84,192 @@ queries rather than introducing a new scalar type. This string is expected to
+One important parameter of FTS is *language*, so we generally assume it will
+be part of the ``index`` declaration to be consistent with the various FTS
+function signatures.
+
+Postgres passes the language as part of the index configuration: ``CREATE
+INDEX pgweb_idx ON pgweb USING GIN (to_tsvector('english', body));``.
+
+Elasticsearch has language-specific analyzers which can be used in the mapping
+corresponding to a given index.
+
+Algolia has an ``indexLanguages`` setting for its indexes.
+
+Typesense does not appear to have any language-specific index settings.
+
+Additionally there's a common need to index prefixes to facilitate
+prefix-based searches (such as for autocomplete). This can be accomplished by
+adding another parameter to the ``abstract index`` definition::
+
+  abstract index fts::textsearch(
+      named only language: str,
+      named only index_prefixes: bool = false
+  )
+
+Setting up a prefix-based index, wouldn't really do much for Postgres.
+
+In Elasticsearch that would translate into ``index_prefixes`` with default
+minimum and maximum prefix values.
+
+In Algolia prefix searches are by default enabled for all, but can be disabled
+via ``disablePrefixOnAttributes``, which would correspond to not turning on
+``index_prefixes``.
+
+In Typesense prefix searching is entirely controlled by the query and no
+special index settings are necessary one way or another, much like in
+Postgres.
+
+Concrete Index
+--------------
+
+As mentioned earlier concrete indexes on types will specify which properties
+the FTS functions should be targeting. When targeting multiple properties they
+can be specified as separate index ``on`` parameters or just concatenated into
+a single expression. Keeping the parameters separate makes it possible to
+assign different weights to them.
+
+For example::
+
+  type Document {
+      title: str;
+      body: str;
+      internal: str;
+
+      index fts::textsearch(language := 'english')
+        on (.title, .body);
   }
 
-If ranking was requested, then the value will be written in ``rank``.
+The above schema declares that only ``title`` and ``body`` properties are
+subject to FTS functions, while the ``internal`` property will be ignored by
+searches.
 
-If highlighting the matched text was requested, then the appropriate value
-will be written in ``highlights`` array. The array is needed in case the
-matches are highlighted separately (e.g. only the matched words are returned,
-omitting the surrounding text).
+In order to add weights, we can introduce a special pseudo-function which can
+only appear in index specification::
 
-These enhanced FTS matches would be performed by the following built-in
-functions::
+  type Document {
+      title: str;
+      body: str;
+      internal: str;
 
-  function fts::match(
-    doc: str,
-    query: str,
-    rank_opts: optional str = 'default',
-    highlight_opts: optional str = {}
-  ) -> FTSResult
+      index fts::textsearch(language := 'english')
```

Comment:
We probably want `fts::weight` and friends to return some opaque internal type. We'll disallow having it appear in non-index queries